### PR TITLE
Take initializing shards into consideration during awareness allocation

### LINKED ISSUE #12522: Inconsistent awareness allocation
When awareness attribute is unbalanced, behavior of shard allocation is inconsistent depending on how many shards are already started.

Reproduction: https://gist.github.com/masaruh/63db5a030220a26cc2a8
It allocates all shards of an index to nodes on creation but when one of nodes leaves and comes back, it doesn't allocate previously allocated shard back to the node anymore.
(All shards shouldn't be allocated in the first place?)

Probably because AwarenessAllocationDecider doesn't take initializing shards into consideration when it checks if shards can be allocated on nodes (https://github.com/elastic/elasticsearch/blob/v1.7.0/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java#L189-L198)?

Looks to be related to #12431.

### PR DESCRIPTION
**Pull request by masaruh**
It makes decision consistent.
Fixes #12522

### CODE REVIEW

**Comment by dakrone** [comment]
This looks good to me, but I think @s1monw should take a look also just to make sure.

Also, not sure about where this should go, @clintongormley what do you think?

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/AwarenessAllocationTests.java [hunk 1]
```diff
@@ -763,7 +763,7 @@ public void fullAwareness3() {
     }
 
     @Test
-    public void testUnbalancedZones() {
+    public void testUnbalancedZones1() {
```
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/AwarenessAllocationTests.java]: can we get a better name for this?

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/AwarenessAllocationTests.java [hunk 2]
```diff
@@ -826,4 +826,45 @@ public void testUnbalancedZones() {
         assertThat(clusterState.getRoutingNodes().node("A-0").size(), equalTo(3));
         assertThat(clusterState.getRoutingNodes().node("B-0").size(), equalTo(5));
     }
+
+    @Test
+    public void testUnbalancedZones2() {
```
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/AwarenessAllocationTests.java]: same here :)

**Comment by s1monw** [comment]
left trivial comments - LGTM though

**Comment by s1monw** [comment]
I think we should let it bake on master and 2.x and if it's stable port to 2.0?

**Comment by masaruh** [comment]
Thanks. Updated test names.
Pushed it to master, 2.x and 1.7.

**Comment by s1monw** [comment]
if you push to 1.7 you also need to push to 2.0 otherwise this makes no sense

**Comment by masaruh** [comment]
oh... pushed to 2.0 as well.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java [hunk 3]
```diff
@@ -188,7 +188,7 @@ private Decision underCapacity(ShardRouting shardRouting, RoutingNode node, Rout
                 if (assignedShard.relocating()) {
                     RoutingNode relocationNode = allocation.routingNodes().node(assignedShard.relocatingNodeId());
                     shardPerAttribute.addTo(relocationNode.node().attributes().get(awarenessAttribute), 1);
-                } else if (assignedShard.started()) {
+                } else if (assignedShard.started() || assignedShard.initializing()) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java]: This means we are counting target relocation shards twice.  Once with the source shards, and once here. We can remove the assignedShard.relocating part yet.

**Comment by bleskes** [comment]
Late to the party, but I think there is a problem with this fix. Left a comment on it. The fact we didn't catch it also means our tests are not strong enough. @masaruh let me know if you want to continue here, o.w. I'll pick it up.

**Comment by masaruh** [comment]
@bleskes you are right... Created #13512 (the test fails without the fix).

**Pull request merged**